### ISSUE TITLE
Questions on supervised traning part code and performance

### ISSUE THREAD

**Issue by Kin-Zhang** [kylevedder/zeroflow#2]
Thanks for your work and open source,

When I read the code following, I'm wondering why set a assert here for z index must be 0? since Voxelization is 3D, z should not be 0?

https://github.com/kylevedder/zeroflow/blob/58a93f7933c7b86122c4d6e6184f612a25ce818d/models/heads/fast_flow_decoder.py#L15-L32

**Comment by Kin-Zhang** [comment]
I see, saw your config to set z voxel size to the range, so for the config, it will be 0 [len=1] but why for that? 2D grid is better?

**Comment by kylevedder** [comment]
The student is the FastFlow3D model, which uses a PointPillars feature encoder that turns everything into a 2D birds eye view pseudoimage.

The voxelization function that I am using is provided by MMCV, which is more general and can be used for 3D voxelization (e.g. SECOND / VoxelNet). I set the minimum and maximum point height in the config as you referenced, and the point clouds should be chopped accordingly, so everything should be in a single very tall voxel (a point _pillar_, hence the name PointPillars) to form the pseudoimage. I added the referenced assert to validate this assumption when doing the voxilization for FastFlow3D; if something had a Z index other than zero, it means that the assumption is being violated and that assert should trigger.

**Comment by Kin-Zhang** [comment]
Thanks for your reply. 🥰

**Comment by Kin-Zhang** [comment]
Sorry to bother you again.

I'm trying directly with ground truth flow on av2 dataset with zeroflow fastflow3d part. But the loss is weird and the eval score is pretty low... 0.699

1. param: lr (2e-6), batch size(16), epoch (50) is exactly the same as your scripts. others are defaulted by the function.
2. loss is weights as 0.9 for dynamic and 0.1 for static.
3. There is no change on model scripts directly using the repo one.

Here is the training loss:
![image](https://github.com/kylevedder/zeroflow/assets/35365764/4dac328a-97af-4509-b923-e3b6d1363133)

And here is the eval for zeroflow after I tried on av2 ground truth flow:
![image](https://github.com/kylevedder/zeroflow/assets/35365764/279b5f8a-fede-4fe0-bd56-31d64ebaf6d0)

Did you also have such experience and maybe some tips for tune?

**Comment by kylevedder** [comment]
Download the pre-trained model weights and run the evaluation on those to ensure that everything else is setup correctly.

If you're able to reproduce those test numbers, then there's something going on with the training run that we can dig into further.

**Comment by Kin-Zhang** [comment]
Thanks for your help! Appreciate.

By the way, I saw there is new eval on zeroflow in leaderboard
1. what's mean on XL?
2. is it still zeroflow pipeline [but it's kind of weird since it is over the teacher NSFP?] or fastflow3d with gt flow to supervise for the result in the leaderboard?

**Comment by kylevedder** [comment]
ZeroFlow XL is the ZeroFlow pipeline with two changes:

1. We use twice as much data, pulling unlabeled point clouds from the Argoverse 2 LiDAR dataset (I updated GETTING_STARTED.md with details)
2. We use an enlarged student model; we quadruple the pseudoimage area (512x512 to 1024x1024 by making the pillars twice as small in each dimension), we double the size of the point embedding vector (thereby making each layer of the UNet twice as fat), and we add another layer to the UNet. To contextualize the size of the model change, the normal student model weights are 79MB, the XL model weights are 1.3GB. I've pushed the [new UNet backbone](https://github.com/kylevedder/zeroflow/commit/47d227c1220ba5b05978e734d1166aa7080c3419) to the repo.

To be clear, like with ZeroFlow, ZeroFlow XL is using _zero_ human labels. Our results are simply because we used more unlabeled data and added more parameters! We are able to beat the teacher model's performance and achieve state-of-the-art on the AV2 test set because our model has seen enough diverse data and is expressive enough to learn to distinguish noise from signal in the teacher labels.

On Sunday when we got this result, I [tweeted](https://twitter.com/KyleVedder/status/1685841442655055872) this cool updated graph showing we are doing better than our scaling laws predicted on the normal student model:

![F2VMqQXXEAAmSyq](https://github.com/kylevedder/zeroflow/assets/6148088/d15f0436-9002-4efc-8e26-779bfcdff6cb)

To further drive home this point, here are the raw results from our submissions to the AV2 Scene Flow test split:

 - [Base ZeroFlow results]( https://evalai.s3.amazonaws.com/media/submission_files/submission_288585/41a52984-fe71-4e32-a5d7-661efb1f340a.json); Threeway EPE of `0.0814`
 - [NSFP results](https://evalai.s3.amazonaws.com/media/submission_files/submission_290680/7a458c56-15fb-4166-b350-4731d1e534e6.txt); Threeway EPE of `0.0684`
 - [ZeroFlow XL results](https://evalai.s3.amazonaws.com/media/submission_files/submission_298502/ca93613a-c051-4f2d-989a-0c2c46a42ac5.json); Threeway EPE of `0.0578`
 
If you look at the linked results, our XL model outperforms the teacher across all three categories of the Threeway EPE, but makes particularly large gains in the static foreground category. This means our model has learned to recognize (a lack of motion) better than NSFP is able to represent because it's seen enough data to know that, in expectation, static objects should have zero flow, even if there's a bit of noise in the teacher labels. This is while, in expectation, also extracting what correct movement vectors look like for moving objects.

I also have more good news! I'm an idiot and forgot to run the XL model with our Speed Scaling feature enabled (Equation 5 from the paper), and so I stopped training this model after only 5 epochs (this is akin to seeing ~10 epochs worth of frames). This means that the XL model is undertrained, _and_ it's missing a feature that provides free Foreground Dynamic EPE improvements (which substantially improves Threeway EPE). We are training a new XL model with these features enabled, and for more epochs, so we should hopefully get even better performance from our new student model.

**Comment by Kin-Zhang** [comment]
thank you so much for sharing these! looking forward to your updates.

**Comment by Kin-Zhang** [comment]
one more question 😊, XL dataset means also needs zeroflow paper pipeline which is NSFP to produce pseudo label first on the LiDAR dataset? or a new pipeline to do that?

**Comment by kylevedder** [comment]
That's correct, I used NSFP to pseudolabel the Argoverse 2 LiDAR dataset data subset. We have a large SLURM cluster with a bunch of old 2080tis so the pseudolabeling only took a few days because I could parallelize across them all.

I used the [`data_prep_scripts/split_nsfp_jobs_sbatch.py`](https://github.com/kylevedder/zeroflow/blob/master/data_prep_scripts/split_nsfp_jobs_sbatch.py) to setup and launch these jobs for both the Sensor and LiDAR subset pseudolabeling.

**Comment by Kin-Zhang** [comment]
@kylevedder  sorry to bother you again, but could you specify which [three models here](https://github.com/kylevedder/zeroflow_weights/tree/master) to get these three results? or maybe one Base Zeroflow results?

[Base ZeroFlow results](https://evalai.s3.amazonaws.com/media/submission_files/submission_288585/41a52984-fe71-4e32-a5d7-661efb1f340a.json); Threeway EPE of 0.0814
[NSFP results](https://evalai.s3.amazonaws.com/media/submission_files/submission_290680/7a458c56-15fb-4166-b350-4731d1e534e6.txt); Threeway EPE of 0.0684
[ZeroFlow XL results](https://evalai.s3.amazonaws.com/media/submission_files/submission_298502/ca93613a-c051-4f2d-989a-0c2c46a42ac5.json); Threeway EPE of 0.0578

**Comment by kylevedder** [comment]
As we discuss in the paper, our reported Threeway EPE for ZeroFlow is an average of three runs.

These weights are the ones highlighted in the weight repo README:

https://github.com/kylevedder/zeroflow_weights/tree/master/argo/nsfp_distilatation_speed_scaled_updated

https://github.com/kylevedder/zeroflow_weights/tree/master/argo/nsfp_distilatation_speed_scaled_updated_run2

https://github.com/kylevedder/zeroflow_weights/tree/master/argo/nsfp_distilatation_speed_scaled_updated_run3

NSFP doesn't have trained weights, it's a test time optimization method.

We have not uploaded the ZeroFlow XL weights, they are too large (1.2GB) and require me to setup git LFS.

**Comment by yanconglin** [comment]
hi, kylevedder,

Could you please let me know the number of samples in the processed Argo/Waymo dataset, train/val/test splits respectively?  It seems there are several versions of waymo scene flow datasets, such as [PCAccumulation, ECCV2022](https://github.com/prs-eth/PCAccumulation). The strategies to calculate gt glows are similar. But I wonder if there is a difference in scale. I can not find any info in the paper and sup. 

It seems ego motion compensation is used when creating the scene flow dataset as mentioned in the sup. Could you please share the results WITHOUT ego-motion compensation, if any.  So far my results show NSFP performs worse on dynamic objects when using ego-motion on waymo. Not sure to what extent this impacts the distillation. Any insights from your side? Thank you!

**Comment by kylevedder** [comment]
**Dataset Details**

For Argoverse2, I read the dataset straight from disk as downloaded, sans the minor folder rearrangement I discuss in the [GETTING_STARTED.md](https://github.com/kylevedder/zeroflow/blob/master/GETTING_STARTED.md#argoverse-2-sensor-dataset).

For Waymo Open, the exact dataset version and labels are detailed in my [GETTING_STARTED.md](https://github.com/kylevedder/zeroflow/blob/master/GETTING_STARTED.md#waymo-open) -- we use 1.4.2 and use the standard flow labels provided on the Waymo Open website. We preprocess the data from the annoying `.proto` format into easy to read `.pkl` files, along the way removing the ground plane from the point clouds. For details, please read the preprocessing scripts discussed in the Getting Started, they are pretty easy to read and frankly I do not remember all the nuances of my preprocessing.

**ZeroFlow without ego motion compensation**

We do not have any results for ZeroFlow / FastFlow3D without ego motion compensation. In principle we can train our feedforward model without ego compensation, but it's reasonable to assume decent quality ego compensation is available at test time on modern service robot / autonomous vehicle stacks. [Chodosh et al 2023](https://arxiv.org/abs/2304.02150) makes a fairly compelling case that ego compensation in general is broadly useful, so we decided to use it. 

**NSFP without ego motion compensation**

I don't directly have head to head NSFP results with and without ego compensation. In my early work using NSFP on Argoverse2 I saw Threeway EPE was better with compensation (which makes sense, it's an easier problem), and we ran with that on Waymo.

How much worse is NSFP on the dynamic bin? Do you have more details on what kinds of dynamic objects it's performing worse on / when? Are you doing ground removal? (this is basically mandatory to get NSFP to work, otherwise it fits a bunch of zero vectors to the lidar on the ground)

I also found that NSFP performance is very dependent upon dataloading details -- ZeroFlow's implementation integrates the author implementation of NSFP, but we use our own data loaders. The NSFP authors actually reached out to me to discuss dataloader details because our NSFP implementation (which is listed as the Baseline NSFP implementation on the [Argoverse2 Scene Flow leaderboard](https://eval.ai/web/challenges/challenge-page/2010/leaderboard/4759)) actually significantly outperformed their own implementation. Their entry is _NP (NSFP)_, my NSFP implementation is the _Host_67820_Team NSFP_ entry (the challenge organizers asked me to send them results for a strong baseline).

**Comment by yanconglin** [comment]
Hi, kylevedder.

This is what I got on the waymo dataset [PCAccumulation, ECCV2022](https://github.com/prs-eth/PCAccumulation):  scene flow estimation between two nearby frames. 
It looks weird that NSFP ego performs worse than NSFP w/o ego on dynamic and the margin is substantially large.

![Screenshot from 2023-10-24 14-27-04](https://github.com/kylevedder/zeroflow/assets/26435516/7fa8ef8b-eac7-44ad-a2ca-152315513ebb)

**Comment by kylevedder** [comment]
Hi Yancong,

Thanks for posting your result table.

I don't know why ego compensated NSFP performs so much worse on your
dynamic objects. My recommendation is you actually visualize the results
for both ego compensated and uncompensated for the same point clouds to
make sure that 1) your data is setup properly 2) NSFP is actually
optimizing the right things 3) NSFP has actually converged

I'm curious to find out what you discover from this, please keep me posted!

**Comment by Kin-Zhang** [comment]
> I also found that NSFP performance is very dependent upon dataloading details -- ZeroFlow's implementation integrates the author implementation of NSFP, but we use our own data loaders.

I also found when I tried to reproduce the FastFlow3D result which is Zeroflow teacher network. However, since I used the official dataloader inside av2-api, and there are some score lost. I will try to find out in the following days.

**Comment by Kin-Zhang** [comment]
> > I also found that NSFP performance is very dependent upon dataloading details -- ZeroFlow's implementation integrates the author implementation of NSFP, but we use our own data loaders.
> 
> I also found when I tried to reproduce the FastFlow3D result which is Zeroflow teacher network. However, since I used the official dataloader inside av2-api, and there are some score lost. I will try to find out in the following days.

Thanks to @kylevedder , he mentioned one of reasons in the email (in case someone after me has same problem, I attached his words here): 
> If your numbers are significantly worse, then something is wrong. If that's the case, my first guess is that you trained on a single GPU using the given config, which is setup to train on 4x GPUs simultaneously and thus has the per GPU batch size set to 16 instead of 64.

Thanks again to Kyle!
